app/merginapi.cpp

```diff
@@ -64,13 +240,213 @@ ProjectList MerginApi::parseProjectsData(const QByteArray &data)
+        boundary = match.captured("boundary");
+    }
+
+    // plus some safety trashold
+    int boundarySize = boundary.length() + 8;
+    QRegularExpression boundaryPattern("(\r\n)?--" + boundary + "\r\n");
+    QRegularExpression headerPattern("Content-Disposition: form-data; name=\"(?P<name>[^'\"]+)\"(; filename=\"(?P<filename>[^\"]+)\")?\r\n(Content-Type: (?P<content_type>.+))?\r\n");
+    QRegularExpression endPattern("(\r\n)?--" + boundary +  "--\r\n");
+
+    QByteArray data;
+    QString dataString;
+    QString activeFilePath;
+
+    while (true) {
+        QByteArray chunk = r->read(CHUNK_SIZE);
+        if (chunk.isEmpty()) {
+            // End of stream - write rest of data to active file
+            if (!activeFilePath.isEmpty()) {
+                QRegularExpressionMatch endMatch = endPattern.match(data);
+                int tillIndex = data.indexOf(endMatch.captured(0));
+                saveFile(data.left(tillIndex), activeFilePath, true);
+                activeFilePath = "";
+            }
+            return;
+        }
+
+        data = data.append(chunk);
+        dataString = QString::fromStdString(data.toStdString());
+        QRegularExpressionMatch boundaryMatch = boundaryPattern.match(dataString);
+
+        while (boundaryMatch.hasMatch()) {
+            if (!activeFilePath.isEmpty()) {
+                int tillIndex = data.indexOf(boundaryMatch.captured(0));
+                saveFile(data.left(tillIndex), activeFilePath, true);
+                activeFilePath = "";
+            }
+
+            // delete previously written data with next boundary part
+            int tillIndex = data.indexOf(boundaryMatch.captured(0)) + boundaryMatch.captured(0).length();
+            data = data.remove(0, tillIndex); // String != QByteArray
+            dataString = QString::fromStdString(data.toStdString());
+
+            QRegularExpressionMatch headerMatch = headerPattern.match(dataString);
+            if (!headerMatch.hasMatch()) {
+                qDebug() << "Received corrupted header";
+                data = data + r->read(CHUNK_SIZE);
+                dataString = QString::fromStdString(data.toStdString());
+            }
+            headerMatch = headerPattern.match(dataString);
+            data = data.remove(0, headerMatch.captured(0).length());
+            dataString = QString::fromStdString(data.toStdString());
+            QString filename = headerMatch.captured("filename");
+
+            activeFilePath = projectDir + "/" + filename;
+            createPathIfNotExists(activeFilePath);
+            boundaryMatch = boundaryPattern.match(dataString);
+        }
+
+        // Write rest of chunk to file
+        if (!activeFilePath.isEmpty()) {
+            saveFile(data.left(data.size() - boundarySize), activeFilePath, true);
+        }
+        data = data.remove(0, data.size() - boundarySize);
+    }
+}
+
+bool MerginApi::saveFile(const QByteArray &data, QString filePath, bool append)
```

Comment:
I see the saveFile() is always called with append=true - do we need that argument?